File: packages/node/motia-example/steps/todo/create-todo.step.ts
import { ApiRouteConfig, Handlers } from '@iii-dev/motia'
import { z } from 'zod'
import { Todo } from './todo.stream'

export const todoSchema = z.object({
  id: z.string(),
  description: z.string(),
  createdAt: z.string(),
  dueDate: z.string().optional(),
  completedAt: z.string().optional(),
})

export const config: ApiRouteConfig = {
  type: 'api',
  name: 'CreateTodo',
  description: 'Create a new todo item',
  flows: ['todo-app'],

  method: 'POST',
  path: '/todo',
  bodySchema: z.object({
    description: z.string(),
    dueDate: z.string().optional(),
  }),
  responseSchema: {
    200: todoSchema,
    400: z.object({ error: z.string() }),
  },
  emits: [],
  virtualEmits: ['todo-created'],
}

export const handler: Handlers['CreateTodo'] = async (req, { logger, streams }) => {
  logger.info('Creating new todo', { body: req.body })

  const { description, dueDate } = req.body
  const todoId = `todo-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`

  if (!description) {
    return { status: 400, body: { error: 'Description is required' } }
  }

  const newTodo: Todo = {
    id: todoId,
    description,
    createdAt: new Date().toISOString(),
    dueDate: dueDate,
    completedAt: undefined,
  }

  const todo = await streams.todo.set('inbox', todoId, newTodo)

  logger.info('Todo created successfully', { todoId })

  return { status: 200, body: todo }
}

File: packages/python/motia-example/steps/todo/create_todo.step.py
"""Create Todo Step."""

import random
import string
from datetime import datetime
from typing import Any

from motia import ApiRequest, ApiResponse, ApiTrigger, FlowContext, StepConfig, Stream, step_wrapper

todo_stream: Stream[dict[str, Any]] = Stream("todo")


config = StepConfig(
    name="CreateTodo",
    description="Create a new todo item",
    flows=["todo-app"],
    triggers=[
        ApiTrigger(path="/todo", method="POST"),
    ],
    emits=[],
)


async def handler(request: ApiRequest[dict[str, Any]], ctx: FlowContext[Any]) -> ApiResponse[Any]:
    """Handle create todo request."""
    ctx.logger.info("Creating new todo", request.body)

    body = request.body or {}
    description = body.get("description")
    due_date = body.get("due_date")

    if not description:
        return ApiResponse(status=400, body={"error": "Description is required"})

    # Generate unique ID
    random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=7))
    todo_id = f"todo-{int(datetime.now().timestamp() * 1000)}-{random_suffix}"

    new_todo: dict[str, Any] = {
        "id": todo_id,
        "description": description,
        "created_at": datetime.now().isoformat(),
        "due_date": due_date,
        "completed_at": None,
    }

    todo = await todo_stream.set("inbox", todo_id, new_todo)

    ctx.logger.info("Todo created successfully", {"todo_id": todo_id})

    return ApiResponse(status=200, body=todo)


step_wrapper(config, __file__, handler)
